Add ADOT Ruby Documentation

NathanielRN opened the pull request:
## Description

Adds ADOT Ruby documentation for how to instrument OpenTelemetry Ruby for compatibility with the AWS X-Ray Service.

Follow the templates for ADOT documentation here: https://github.com/willarmiros/aws-otel-docs/tree/main/templates

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,362 @@
+  }
+end
+```
+
+Alternatively, once you've downloaded the `Instrumentation` gem, you can simply tell the OpenTelemetry SDK to trace all `Instrumentation`s including the `OpenTelemetry::Instrumentation::AwsSdk` Instrumentation.
+
+```ruby lineNumbers=true
+OpenTelemetry::SDK.configure do |c|
+  c.use_all()
+end
+```
````
```suggestion
```

Since this isn't specifically related to the AWS SDK, and it's mentioned above, let's remove.

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,362 @@
+
+<SectionSeparator />
+
+## Setting up the Global Tracer
```
Sorry for missing this initially - I added a new section for configuring sampling since that's been a pain point for existing customers: https://github.com/willarmiros/aws-otel-docs/blob/main/templates/manual-trace-instrumentation.md#configuring-sampling

If we can let's include this sampling configuration too.

NathanielRN replied:
Cool, I investigated what it would take to configure it with environment variables and with code and added it as a section here!

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,362 @@
+
+With the exporter created, initialize OpenTelemetry Ruby SDK to use it:
+
+```ruby
+OpenTelemetry::SDK.configure do |c|
+  c.span_processors = [span_processor]
+end
+```
````
Let's remove these individual code snippets in this section.

NathanielRN replied:
Removed!

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,362 @@
+Putting this all together we have something like the following:
+
+```ruby
+OpenTelemetry::SDK.configure do |c|
+  c.service_name = 'aws-otel-manual-rails-sample'
+
+  c.span_processors = [
+      Trace::Export::BatchSpanProcessor.new(
+          OpenTelemetry::Exporter::OTLP::Exporter.new(endpoint="http://localhost:4318")
+      )
+    ]
+  c.id_generator = OpenTelemetry::Propagator::XRay::IDGenerator
+  c.propagators = [OpenTelemetry::Propagator::XRay::TextMapPropagator.new]
+end
````
```suggestion
OpenTelemetry::SDK.configure do |c|
  c.service_name = 'aws-otel-manual-rails-sample'

  # Comment explaining batch span processor
  c.span_processors = [
      Trace::Export::BatchSpanProcessor.new(
          OpenTelemetry::Exporter::OTLP::Exporter.new(endpoint="http://localhost:4318")
      )
    ]
    
  # Comment explaining Id generator
  c.id_generator = OpenTelemetry::Propagator::XRay::IDGenerator
  
  # Comment on propagator
  c.propagators = [OpenTelemetry::Propagator::XRay::TextMapPropagator.new]
end
```

Instead, we can add some comments in the larger code snippet to explain.

NathanielRN replied:
Added comments.

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,362 @@
+require 'opentelemetry-propagator-xray'
+```
+
+Next, configure the Global Tracer Provider. This involves configuring
````
Wow, this section has a lot of detail! I think as a general rule of thumb for documentation, we can assume 2 things: 
1. Most customers do not want to read a ton, they just want to get stuff working
2. Some customers will want to get in the weeds and understand very specific things, and those will be the ones who deeply read the docs.

With that in mind, I think we can simplify this section by first having a brief description of what the customer should do, then put the large code snippet that "puts everything together" first with comments that also explain each line. Then we can remove the individual code snippets. The additional details, like environment variable configuration and TLS stuff, can go after the main code snippet for those power users.

NathanielRN replied:
Done!

willarmiros reviewed (commented):
Looks great, mostly small comments and some notes about simplification. Thank you!

NathanielRN commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,362 @@
+
+In a Ruby on Rails app, you will not need to require packages in your application code because of [autoloading](https://guides.rubyonrails.org/getting_started.html#autoloading). This assumes you are using `bundler` and a Gemfile.
+
+However, if you tracing a normal Ruby program, or have included gems with the `, require: false` option, the code below shows step by step how to configure the OpenTelemetry Ruby SDK in any normal Ruby program.
```
Sure I can split it up. For reference, [the OTel Ruby docs](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) talk about Rails in their main documentation section which signal high importance to me. But I'll split it up regardless

NathanielRN commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,362 @@
+
+### Sending Traces to AWS X-Ray
+
+For a ruby on rails application, OpenTelemetry Ruby [recommends using a Rails initializer](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) to configure the OpenTelemetry Ruby SDK. ADOT provides a [working example of such an intializer in our sample app repo](https://github.com/aws-observability/aws-otel-ruby/blob/main/sample-apps/manual-instrumentation/ruby-on- ...
```
I am also unfamiliar with Railtie, but it has come up when I ask OTel Ruby SIG members question and I noticed a SIG member mentioned it to me here: https://github.com/open-telemetry/opentelemetry-ruby/issues/1089#issuecomment-1015914979

There's less evidence than I thought originally to support it being a canonical thing in Ruby, so I will remove it! Especially since we don't have any concrete examples.

NathanielRN commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,362 @@
+end
+```
+
+### Using the AWS resource Detectors
````
Removed!

NathanielRN commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,362 @@
+# Get a tracer from the Global Tracer Provider
+tracer = OpenTelemetry.tracer_provider.tracer('my-tracer')
+
+tracer.in_span('Root Span') do |root_span|
```
Updated! Although on a side note, I thought we changed that last year so that it didn't need to be of type `Server`? Either way I have updated it.

NathanielRN commented:
@willarmiros Addressed your comments! Hopefully you can look at [the latest commit](https://github.com/aws-otel/aws-otel.github.io/pull/224/commits/f184565d6746ff2e25a526d5e0a3a9587d52a2ac) to see just what changed? Thanks for your helpful feedback! 🙂

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -69,13 +69,27 @@ $ gem install opentelemetry-instrumentation-faraday -v '~> 0.19' \
+
+This section describes recommended configuration needed for OpenTelemetry X-Ray tracing with X-Ray.
+
+##### Quick Setup
```
Just want to confirm - does the "quick setup" and "advanced setup" sections only apply to non-rails applications? If not, they shouldn't be nested like this.

NathanielRN replied:
"Quick Setup" and "Advanced Setup" applies to _all_ applications, not just Rails applications.

Hm this is how I figured to split the section into "Rails" vs "non-Rails" as we mentioned previously. Would this split seem better?:
* Setting up the Global Tracer
  * Sending Traces to AWS X-Ray
    * Configure the OpenTelemetry Ruby SDK
    * Tips for Rails Applications

That's why I originally put "rails" in the "configure OTel Ruby SDK" section, because you still have to configure OTel Ruby, the bit about "Rails Initializers" is just a helpful tip. I guess this is [why OTel Ruby Getting Started Docs do not split Rails into its own section](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) but also just keep it together in the "General Initialization Section".

Let me know what you think? I think the "additional note" is a good compromise. Thank you!

willarmiros replied:
Discussed offline, seems like we shouldn't have separate sections for rails vs. non-rails. We can structure it like:

* Setting up the Global Tracer
  * Sending Traces to AWS X-Ray (Includes a quick note about using rails initializers and has the "quick setup" content)
  * Configure the OpenTelemetry Ruby SDK (has the advanced setup content)

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,371 @@
+
+In a Ruby on Rails app, you will not need to require packages in your application code because of [autoloading](https://guides.rubyonrails.org/getting_started.html#autoloading). This assumes you are using `bundler` and a Gemfile.
+
+#### From a normal Ruby Program
```
```suggestion
#### From other Ruby Programs
```

This might confuse Rails users, who probably consider their programs "normal" ;)

NathanielRN replied:
Lol, I must be dizzy or something, I could've sworn "normal" is what the OTel Ruby docs said, but [they say "other" too](https://opentelemetry.io/docs/instrumentation/ruby/getting_started/#initialization) so thanks for htis!

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,371 @@
+
+
+```bash
+OTEL_TRACES_SAMPLER=traceidratio && OTEL_TRACES_SAMPLER_ARG=0.10
````
```suggestion
export OTEL_TRACES_SAMPLER=traceidratio
export OTEL_TRACES_SAMPLER_ARG=0.10
```

willarmiros replied:
Also, just want to confirm this is the `parentbased` sampler right?

NathanielRN replied:
Oh sorry no it is not! I will fix that right away.

NathanielRN replied:
```suggestion
OTEL_TRACES_SAMPLER=parentbased_traceidratio && OTEL_TRACES_SAMPLER_ARG=0.10
```

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,371 @@
+
+By default, the OpenTelemetry Ruby SDK follows the parent span's sampling decision if it exists, and samples 100% of incoming requests otherwise. This is known as the `parentbased_always_on` sampler.
+
+To reduce the sampling rate, configure OpenTelemetry Ruby SDK to use the `parentbased_traceidratio` sampler. This can be configured using [the OpenTelemetry Specification defined environment variables](https://github.com/open-telemetry/opentelemetry-specification/blob/7b504383f53b02b10f62ef78fa008fdfd18c633e/specification/sdk-environment-variables.md#general-sdk-configuration):
```
```suggestion
To reduce the sampling rate, configure OpenTelemetry Ruby SDK to use the `parentbased_traceidratio` sampler. This can be configured using [the OpenTelemetry Specification defined environment variables](https://github.com/open-telemetry/opentelemetry-specification/blob/7b504383f53b02b10f62ef78fa008fdfd18c633e/specification/sdk-environment-variables.md#general-sdk-configuration). For instance, to reduce the sampling rate to 10% of requests, set the following environment variables:
```

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,371 @@
+Alternatively, this can be set on the global TracerProvider after the OpenTelemetry Ruby SDK has been configured:
+
+```ruby
+OpenTelemetry.tracer_provider.sampler = Samplers.trace_id_ratio_based(0.10)
````
Is this `parentbased`?

NathanielRN replied:
Ah no it is not, will update right away.

NathanielRN replied:
```suggestion
OpenTelemetry.tracer_provider.sampler = Samplers.parent_based(root: Samplers.trace_id_ratio_based(0.10))
```

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,371 @@
+By default, OpenTelemetry Ruby SDK [is already configured to initialize an OTLP exporter](https://github.com/open-telemetry/opentelemetry-ruby/tree/main/exporter/otlp#how-do-i-get-started). The exporter can also be completely [configured using environment variables](https://github.com/open-telemetry/opentelemetry-ruby/tree/main/exporter/otlp#how-can-i-configure-the-otlp-exporter).
+
+```bash
+OTEL_TRACES_EXPORTER=otlp && OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
````
```suggestion
export OTEL_TRACES_EXPORTER=otlp
export OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
```

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,371 @@
+
+Next, because the AWS X-Ray ID Generator can only be configured through code, you cannot use an environment variable to select it. The ID Generator must be used at the time OpenTelemetry Ruby SDK is configured.
+
+Finally,tTo allow the span context to propagate downstream when the application makes calls to external services, configure the global propagator to use the AWS X-Ray Propagator.
```
```suggestion
Finally, to allow the span context to propagate downstream when the application makes calls to external services, configure the global propagator to use the AWS X-Ray Propagator.
```

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,371 @@
+
+##### Quick Setup
+
+When you instrument a normal Ruby program, or if you included the gems with the `require: false` option, you will need to "require" the gems distributed by OpenTelemetry manually.
```
```suggestion
When you instrument a non-rails Ruby program, or if you included the gems with the `require: false` option, you will need to "require" the gems distributed by OpenTelemetry manually.
```

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -292,9 +307,12 @@ require 'opentelemetry-api'
 tracer = OpenTelemetry.tracer_provider.tracer('my-tracer')
 
-tracer.in_span('Root Span') do |root_span|
+tracer.in_span('Root Span', kind: :server) do |root_span|
```
Is the space between colons intentional? If so, man is Ruby weird

NathanielRN replied:
Lol yes it is, you can see [an example on the OTel Ruby Getting Starter Docs - Manually Creating your own spans](https://opentelemetry.io/docs/instrumentation/ruby/manual_instrumentation/#creating-new-spans)

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,371 @@
+
+tracer.in_span('Root Span', kind: :server) do |root_span|
+
+    p 'Started a root span'
```
```suggestion
    p 'Started a root span, this will be a segment in the X-Ray console'
```

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,371 @@
+
+    tracer.in_span('Child Span') do |child_span|
+
+        p 'Started a child span'
```
```suggestion
        p 'Started a child span, this will be a subsegment in the X-Ray console'
```

willarmiros reviewed (commented):
Looks great! All really small suggestions, then will be good to go.

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
````diff
@@ -0,0 +1,371 @@
+end
+```
+
+With this, your Ruby application has configured OpenTelemetry Ruby to trace for the X-Ray backend service! To automatically trace popular Ruby gems, jump to the next section to learn about [instrumenting with OpenTelemetry Ruby Instrumentations](#instrumenting-an-application).
````
Yep it should!

willarmiros commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,371 @@
+
+This section describes recommended configuration needed for OpenTelemetry X-Ray tracing with X-Ray.
+
+##### Quick Setup
```
Should be ok

willarmiros reviewed (approved):
Thanks for all the changes!

alolita reviewed (approved):
Readability improvements added. Lgtm.

NathanielRN commented on src/docs/getting-started/ruby-sdk/trace-manual-instr.mdx:
```diff
@@ -0,0 +1,370 @@
+
+With OpenTelemetry Ruby manual instrumentation, you configure the OpenTelemetry SDK within your application with just a few lines of code. OpenTelemetry Ruby then automatically produces trace spans with telemetry data describing the values used by the Ruby gems in your application. This telemetry data can then be exported to a backend like AWS X-Ray using the `OpenTelemetry::Propagator::XRay::IDG ...
+
+In this guide, we walkthrough the steps needed to trace an application with manual instrumentation.
```
We should use 2 words for this part here

```suggestion
In this guide, we walk through the steps needed to trace an application with manual instrumentation.
```

alolita commented:
@NathanielRN can you check and fix this PR check is failing for one of the tests.

NathanielRN commented:
@alolita Looks like it was an intermittent connection issue 😞

```
error UNHANDLED EXCEPTION read ECONNRESET
```

Can you try re-running workflows?

alolita commented:
@NathanielRN you may need to push a commit to re-trigger the workflow.
